rapid/src/main/java/com/vrg/rapid/MembershipView.java

```diff
@@ -46,6 +49,8 @@
     @GuardedBy("rwLock") private final ArrayList<Utils.AddressComparator> addressComparators;
     @GuardedBy("rwLock") private final ArrayList<NavigableSet<Endpoint>> rings;
     @GuardedBy("rwLock") private final Set<NodeId> identifiersSeen = new TreeSet<>(NodeIdComparator.INSTANCE);
+    @GuardedBy("rwLock") private final Map<Endpoint, List<Endpoint>> cachedObservers = new HashMap<>();
```

Comment:
I see three main changes here:

The first is:
1. Some, but not all uses of `rings.get(0).contains()` have been replaced with `allNodes.contains()`. Just curious: what was the performance difference here (I can't tell what the raw numbers are from the flame graph)?

The next two both have to do with improving the performance of getObserversOf()

 2. ringAdd() and ringDelete() involve updating the cached set of observers for a given subject.
 3. Caching the results of getObserversOf().

I think change 3 alone should suffice here. Change 2 seems heavy: when a batch of membership changes come in, it might invalidate the cache for the same subject several times (especially in smaller clusters).

Reply:
1) `TreeSet.contains` was taking 11% of CPU time:

![treeset-contains](https://user-images.githubusercontent.com/459669/77646856-a2af5100-6f65-11ea-8b86-7e80dba5b58e.png)

Versus, well, nothing noticeable anymore after switching to a `HashSet` (look for traces that start with `com.vrg`, it's really not much anymore):

![after-perf](https://user-images.githubusercontent.com/459669/77647085-16515e00-6f66-11ea-9acb-fae07ca3824e.png)

2) Indeed, invalidating - rather than recomputing - should be sufficient here. Will make the change when I get a chance to do so!

Reply:
Re 1.: Very nice change indeed! Thanks!